### ISSUE TITLE
[Regression] Games have crackling sound with FAudio

### ISSUE THREAD

**Issue by FinishedFragment** [RPCS3/rpcs3#11209]
## Quick summary

The sound is crackling. This happens right after starting the game, too and does not even require beginning a new game or loading a save.

## Details

I built the latest master and noticed the crackling. I am currently on 71b15c1f and it does not have this issue.

#### 0. Make sure you're running with settings as close to default as possible

Write color buffers is enabled, as it is required. Async shader compilation is enabled, too but this should not be the issue.

#### 1. Please provide _exact_ build (or commit) information that introduced the regression you're reporting.

Not sure about the exact commit. Might bisect it later, but it happened between 71b15c1f and 4fe17fcec.

#### 2. Please attach TWO logs:

* One for the build with regression. [RPCS3-regressed.log](https://github.com/RPCS3/rpcs3/files/7613346/RPCS3-regressed.log)
* One for the build that works as expected. [RPCS3.log](https://github.com/RPCS3/rpcs3/files/7613342/RPCS3.log)

#### 3. If you describe graphical regression, please provide an RSX capture and a RenderDoc capture that demonstrate it.

n/a

#### 4. Please attach screenshots of your problem.

n/a, as this is sound. fps are nice and steady, so it does not appear to be related to performance.

#### 5. Please provide comparison with real PS3.

n/a, as this is a regression.

#### 6. Please provide your system configuration:

Arch Linux, rx 570, 8 GB of RAM, Ryzen 1600af, PulseAudio.

**Comment by nitroos-ai** [comment]
This #11081 might be the culprit.

**Comment by GitArUs** [comment]
@FinishedFragment: Have You tried setting audio output to cubeb instead of FAudio ?

**Comment by dariox86** [comment]
I am on revision f910a15, same operating system, same issue with The Last of Us (BCES01584). Setting to Cubeb fixes the issue.

**Comment by FinishedFragment** [comment]
@GitArUs No I did not. I just tried it and the crackling vanished. Thank you very much for that suggestion.

**Comment by GitArUs** [comment]
It probably means, that buffering does not work correctly for FAudio after mentioned PR was merged.

**Comment by AniLeo** [comment]
Missing confirmation of the PR that broke FAudio, presumably 11081, by testing the previous build with FAudio as well and seeing if it works
https://rpcs3.net/builds

**Comment by Megamouse** [comment]
try with #12186

**Comment by FinishedFragment** [comment]
a bit late but since said PR got merged and I am using the latest build, it is still not fixed. while testing out things for #12415, i noticed that the crackling is gone when you start the game with cubeb and switch to faudio in the game-specific config and apply this change.

**Comment by FinishedFragment** [comment]
Audio-only mp4 of the crackling:

https://user-images.githubusercontent.com/44732327/192716025-0fdb2e37-c6b7-4811-9086-1a0deb18d823.mp4

I will try some old appimages now.

**Comment by FinishedFragment** [comment]
Bisecting done! 107107107c1aedd1c7a8edc16843533a26808522 (#11358) appears to be the culprit. Pinging @Vestrel.

This was super awkward to bisect. I trust past me that it did not work for them, but the appimages of this time frame all worked for me. This is not the first time soundtouch acted up and my guess is that my own build used a newer soundtouch version because I might have used my system version. The appimage did probably not until that PR, but these are just my guesses.
The PR looks very relevant to this in any case.

**Comment by kd-11** [comment]
I should mention, I also have this issue and it affects both Cubeb and Faudio. It can be worked around by enabling timestretch but that is more of a workaround than a proper solution. I didn't investigate much further than that.

**Comment by FinishedFragment** [comment]
I can reproduce with both on the latest builds. Maybe some library or submodule change did not propagate to me yet at that time. Maybe an upstream regression? I also think playing around with the latency changes the behavior. I forgot what I did exactly though.

Other projects using either FAudio or Cubeb are not affected for me. So it is rpcs3 specific after all.

**Comment by FinishedFragment** [comment]
I can not eliminate the crackling on FAudio by playing with the latency but lowering the latency on Cubeb (from 100 to 20 ms or so) got rid of the crackling.

**Comment by mvf** [comment]
This also affects Cubeb, so maybe the "with FAudio" should be removed from the issue title.

> It can be worked around by enabling timestretch [...]

This works for me too.

> [...] lowering the latency on Cubeb (from 100 to 20 ms or so) got rid of the crackling.

This does reduce the crackling a lot, but in my setup it's not always enough.

> Bisecting done! 1071071 (#11358) appears to be the culprit.

Can confirm. 0.0.20-13245 AppImage is good, 0.0.20-13248 AppImage is bad.

----

107107107c1aedd1c7a8edc16843533a26808522 changed the audio ringbuffer size calculation in [rpcs3/Emu/Cell/Modules/cellAudio.cpp](https://github.com/RPCS3/rpcs3/blob/bb2c9196f59d19f045809390863c3ec84382a4af/rpcs3/Emu/Cell/Modules/cellAudio.cpp#L149):
```diff
 	const f64 buffer_dur_mult = [&]()
 	{
-		const f64 min_buf_dur = _cfg.audio_min_buffer_duration + 0.01; // Add 10ms to allow jitter compensation
-		if (cfg.raw.buffering_enabled)
+		if (cfg.buffering_enabled)
 		{
-			return std::max<f64>(min_buf_dur, cfg.raw.desired_buffer_duration / 1000.0 * 2); // Allocate 2x buffer to keep buffering algorithm happy
+			return cfg.desired_buffer_duration / 1'000'000.0 + 0.02; // Add 20ms to buffer to keep buffering algorithm happy
 		}
 
-		return min_buf_dur;
+		return cfg.audio_min_buffer_duration;
 	}();
	[...]
	cb_ringbuf.set_buf_size(static_cast<u32>(cfg.audio_channels * cfg.audio_sampling_rate * cfg.audio_sample_size * buffer_dur_mult));
```

For instance, with a 100ms buffer duration, on 3a804674c91bfa75a2a4f7da7f624af801be1eb8 `buffer_dur_mult` was 0.2 (100ms * 2). On 107107107c1aedd1c7a8edc16843533a26808522, it's 0.12 (100ms + 20ms). The reduced size seems insufficient with time stretch disabled. Lowering the buffer duration in the settings can improve things by bringing the effective buffer size closer to what it was before the change.

----

I'm running bb2c9196f59d19f045809390863c3ec84382a4af + this patch for now:
```diff
 	{
 		if (cfg.buffering_enabled)
 		{
-			return cfg.desired_buffer_duration / 1'000'000.0 + 0.02; // Add 20ms to buffer to keep buffering algorithm happy
+			return cfg.desired_buffer_duration / 1'000'000.0 + 0.03; // Add 30ms to buffer to keep buffering algorithm happy
 		}
 
 		return cfg.audio_min_buffer_duration;
```

In my setup, the additional 10ms fix Cubeb with a buffer duration from 4 to ~60ms. Above that, crackling gradually sets in again, the opposite of what one would expect. Maybe someone familiar with the code has an idea.

**Comment by kd-11** [comment]
What about just reverting back to 2x size? Does that fix it?

**Comment by Megamouse** [comment]
Closing due to inactivity. superseded by #13310
